applications/FluidDynamicsApplication/custom_utilities/cfd_utilities.cpp

```diff
@@ -0,0 +1,429 @@
+    const double shear_stress = norm_2(rFrictionVelocity);
+
+    // calculate y_plus
+    const double y_plus = std::sqrt(shear_stress / Density) * WallHeight / KinematicViscosity;
+
+    // calculate u_tau
+    noalias(rFrictionVelocity) =
+        rFrictionVelocity /
+        std::sqrt((shear_stress <= std::numeric_limits<double>::epsilon() ? 1.0 : shear_stress) *
+                  Density);
+
+    return y_plus;
+
+    KRATOS_CATCH("");
+}
+
+double CalculateLinearLogarithmicWallFunctionBasedYPlusLimit(const double VonKarman,
+                                                             const double WallSmoothness,
+                                                             const int MaxIterations,
+                                                             const double Tolerance)
+{
+    double y_plus = 11.06;
+    const double inv_kappa = 1.0 / VonKarman;
+    double dx = 0.0;
+    for (int i = 0; i < MaxIterations; ++i)
+    {
+        const double value = inv_kappa * std::log(y_plus) + WallSmoothness;
+        dx = value - y_plus;
+
+        if (std::abs(dx) < Tolerance)
+            return y_plus;
+
+        y_plus = value;
+    }
+
+    KRATOS_WARNING("CalculateLinearLogarithmicWallFunctionBasedYPlusLimit")
+        << "Logarithmic y_plus limit reached max iterations with dx > "
+           "Tolerance [ "
+        << dx << " > " << Tolerance << ", MaxIterations = " << MaxIterations << " ].\n";
+    return y_plus;
+}
+
+double CalculateLinearLogarithmicWallFunctionBasedYPlusAndUtau(
+    array_1d<double, 3>& rFrictionVelocity,
+    const array_1d<double, 3>& rWallVelocity,
+    const array_1d<double, 3>& rNormal,
+    const double KinematicViscosity,
+    const double WallHeight,
+    const double VonKarman,
+    const double WallSmoothness,
+    const int MaxIterations,
+    const double Tolerance)
+{
+    KRATOS_TRY
+
+    // calculating unit normal since rNormal contains surface area of the condition
+    const double surface_area = norm_2(rNormal);
+    const array_1d<double, 3>& unit_normal = rNormal / surface_area;
+
+    // calculate tangential velocity
+    noalias(rFrictionVelocity) =
+        rWallVelocity - unit_normal * inner_prod(rWallVelocity, unit_normal);
+    const double wall_velocity = norm_2(rFrictionVelocity);
+
+    // calculate linear log region limit
+    const double limit_y_plus = CalculateLinearLogarithmicWallFunctionBasedYPlusLimit(
+        VonKarman, WallSmoothness, MaxIterations, Tolerance);
+
+    // linear region
+    double u_tau = std::sqrt(wall_velocity * KinematicViscosity / WallHeight);
+    double y_plus = u_tau * WallHeight / KinematicViscosity;
+    const double inv_kappa = 1.0 / VonKarman;
+
+    // log region
+    if (y_plus > limit_y_plus)
+    {
+        int iter = 0;
+        double dx = 1e10;
+        double u_plus = inv_kappa * std::log(y_plus) + WallSmoothness;
+
+        while (iter < MaxIterations && std::fabs(dx) > Tolerance * u_tau)
```

Comment:
Why `std::fabs` if `dx` is already  a float?

Reply:
copy paste error :)